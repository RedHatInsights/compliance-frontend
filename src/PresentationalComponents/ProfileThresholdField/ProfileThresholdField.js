import React from 'react';
import { Field, reduxForm } from 'redux-form';
import { FormGroup } from '@patternfly/react-core';
import { ReduxFormTextInput } from 'PresentationalComponents/ReduxFormWrappers/ReduxFormWrappers';
import propTypes from 'prop-types';
import round from 'lodash/round';
import { thresholdValid } from '../../SmartComponents/CreatePolicy/validate';
import { PolicyThresholdTooltip } from 'PresentationalComponents';

export class ProfileThresholdField extends React.Component {
    state = {
        validThreshold: thresholdValid(this.props.previousThreshold),
        threshold: round(this.props.previousThreshold || 100, 1)
    };

    handleThresholdChange = (threshold) => (
        this.setState({
            validThreshold: thresholdValid(threshold),
            threshold: round(threshold, 1)
        })
    )

    render() {
        const { threshold, validThreshold } = this.state;

        return (
            <React.Fragment>
                <FormGroup fieldId='policy-threshold'
                    validated={ validThreshold ? 'default' : 'error' }
                    helperTextInvalid='Threshold has to be a number between 0 and 100'
                    helperText="A value of 95% or higher is recommended"
                    labelIcon={ <PolicyThresholdTooltip /> }
                    label='Compliance threshold (%)'>
                    <Field name='complianceThreshold' id='complianceThreshold' isRequired={true}
                        onChange={this.handleThresholdChange}
                        validated={ validThreshold ? 'default' : 'error' }
                        defaultValue={threshold}
                        aria-label="compliance threshold"
                        component={ReduxFormTextInput} type='number'
                        style={ { width: '60%', display: 'block' } } />
                </FormGroup>
            </React.Fragment>
        );
    }
}

ProfileThresholdField.propTypes = {
    previousThreshold: propTypes.number
};

export default reduxForm({
    form: 'policyForm',
    destroyOnUnmount: false,
    forceUnregisterOnUnmount: true
})(ProfileThresholdField);
