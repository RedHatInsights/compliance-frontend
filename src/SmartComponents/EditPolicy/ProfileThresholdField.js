import React from 'react';
import { Field, reduxForm } from 'redux-form';
import { FormGroup, Title } from '@patternfly/react-core';
import { ReduxFormTextInput } from 'PresentationalComponents/ReduxFormWrappers/ReduxFormWrappers';
import propTypes from 'prop-types';

class ProfileThresholdField extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            validThreshold: true,
            threshold: props.previousThreshold
        };
    }

    handleThresholdChange = threshold => {
        if (threshold > 100 || threshold < 0) {
            this.setState({ validThreshold: false });
        } else {
            this.setState({ validThreshold: true, threshold });
        }
    };

    render() {
        const { threshold, validThreshold } = this.state;
        return (
            <React.Fragment>
                <Title headingLevel="h3" size="xl">Compliance threshold</Title>
                The compliance threshold defines what percentage of rules must be met in order for a system to
                be determined &quot;compliant&quot;.
                <FormGroup field-id='policy-threshold'
                    isValid={validThreshold}
                    helperTextInvalid='Threshold has to be a number between 0 and 100'
                    helperText="A value of 95% or higher is recommended"
                    label="Compliance threshold (%)">
                    <Field name='complianceThreshold' id='complianceThreshold' isRequired={true}
                        onChange={this.handleThresholdChange}
                        isValid={validThreshold}
                        defaultValue={threshold}
                        aria-label="compliance threshold"
                        component={ReduxFormTextInput} type='number'
                        style={ { width: '60%', display: 'block' } } />
                </FormGroup>
            </React.Fragment>
        );
    }
}

ProfileThresholdField.propTypes = {
    previousThreshold: propTypes.number
};

export default reduxForm({
    form: 'editPolicy'
})(ProfileThresholdField);
