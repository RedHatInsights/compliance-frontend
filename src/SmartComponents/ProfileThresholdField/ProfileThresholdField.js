import React from 'react';
import { Field, reduxForm } from 'redux-form';
import { FormGroup, Title, Popover, PopoverPosition } from '@patternfly/react-core';
import { OutlinedQuestionCircleIcon } from '@patternfly/react-icons';
import { ReduxFormTextInput } from 'PresentationalComponents/ReduxFormWrappers/ReduxFormWrappers';
import propTypes from 'prop-types';
import round from 'lodash/round';
import { thresholdValid } from '../CreatePolicy/validate';

export class ProfileThresholdField extends React.Component {
    state = {
        validThreshold: thresholdValid(this.props.previousThreshold),
        threshold: round(this.props.previousThreshold || 100, 1)
    };

    handleThresholdChange = (threshold) => (
        this.setState({
            validThreshold: thresholdValid(threshold),
            threshold: round(threshold, 1)
        })
    )

    render() {
        const { threshold, validThreshold } = this.state;
        const { showTitle } = this.props;
        const explanation = <React.Fragment>The compliance threshold defines what percentage of rules
            must be met in order for a system to be determined &quot;compliant&quot;.</React.Fragment>;
        const titleHeader = <Title headingLevel="h3" size="xl">Compliance threshold</Title>;
        const title = <React.Fragment>
            { titleHeader }
            { explanation }
        </React.Fragment>;
        const popover = <Popover
            position={PopoverPosition.top}
            headerContent={titleHeader}
            bodyContent={explanation}
        >
            <OutlinedQuestionCircleIcon />
        </Popover>;

        return (
            <React.Fragment>
                { showTitle && title }
                <FormGroup fieldId='policy-threshold'
                    validated={ validThreshold ? 'default' : 'error' }
                    helperTextInvalid='Threshold has to be a number between 0 and 100'
                    helperText="A value of 95% or higher is recommended"
                    label={<React.Fragment>Compliance threshold (%) {!showTitle && popover}</React.Fragment>}>
                    <Field name='complianceThreshold' id='complianceThreshold' isRequired={true}
                        onChange={this.handleThresholdChange}
                        validated={ validThreshold ? 'default' : 'error' }
                        defaultValue={threshold}
                        aria-label="compliance threshold"
                        component={ReduxFormTextInput} type='number'
                        style={ { width: '60%', display: 'block' } } />
                </FormGroup>
            </React.Fragment>
        );
    }
}

ProfileThresholdField.propTypes = {
    previousThreshold: propTypes.number,
    showTitle: propTypes.bool
};

export default reduxForm({
    form: 'policyForm',
    destroyOnUnmount: false,
    forceUnregisterOnUnmount: true
})(ProfileThresholdField);
