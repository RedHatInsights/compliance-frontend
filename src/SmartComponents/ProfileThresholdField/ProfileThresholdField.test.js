import toJson from 'enzyme-to-json';

import { ProfileThresholdField } from './ProfileThresholdField';

describe('ProfileThresholdField', () => {
    const defaultProps = {
        previousThreshold: 10
    };

    it('expect to render without error', () => {
        const wrapper = shallow(
            <ProfileThresholdField { ...defaultProps }/>
        );

        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('expect to render without title', () => {
        const wrapper = shallow(
            <ProfileThresholdField showTitle={false} { ...defaultProps }/>
        );

        expect(toJson(wrapper)).toMatchSnapshot();
    });
});
