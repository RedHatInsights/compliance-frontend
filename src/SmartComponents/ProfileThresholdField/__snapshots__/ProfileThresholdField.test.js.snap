// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfileThresholdField expect to handle ThresholdChange 1`] = `
Object {
  "threshold": 100,
  "validThreshold": true,
}
`;

exports[`ProfileThresholdField expect to handle ThresholdChange 2`] = `
Object {
  "threshold": 50,
  "validThreshold": true,
}
`;

exports[`ProfileThresholdField expect to handle ThresholdChange 3`] = `
Object {
  "threshold": 0,
  "validThreshold": false,
}
`;

exports[`ProfileThresholdField expect to render with title 1`] = `
<Fragment>
  <Title
    headingLevel="h3"
    size="xl"
  >
    Compliance threshold
  </Title>
  The compliance threshold defines what percentage of rules must be met in order for a system to be determined "compliant".
  <FormGroup
    fieldId="policy-threshold"
    helperText="A value of 95% or higher is recommended"
    helperTextInvalid="Threshold has to be a number between 0 and 100"
    isValid={true}
    label={
      <React.Fragment>
        Compliance threshold (%) 
      </React.Fragment>
    }
  >
    <Field
      aria-label="compliance threshold"
      component={[Function]}
      defaultValue={10}
      id="complianceThreshold"
      isRequired={true}
      isValid={true}
      name="complianceThreshold"
      onChange={[Function]}
      style={
        Object {
          "display": "block",
          "width": "60%",
        }
      }
      type="number"
    />
  </FormGroup>
</Fragment>
`;

exports[`ProfileThresholdField expect to render without error 1`] = `
<Fragment>
  <FormGroup
    fieldId="policy-threshold"
    helperText="A value of 95% or higher is recommended"
    helperTextInvalid="Threshold has to be a number between 0 and 100"
    isValid={true}
    label={
      <React.Fragment>
        Compliance threshold (%) 
        <Popover
          appendTo={[Function]}
          aria-label=""
          bodyContent={
            <React.Fragment>
              The compliance threshold defines what percentage of rules must be met in order for a system to be determined "compliant".
            </React.Fragment>
          }
          boundary="window"
          className=""
          closeBtnAriaLabel="Close"
          distance={25}
          enableFlip={true}
          flipBehavior={
            Array [
              "top",
              "right",
              "bottom",
              "left",
              "top",
              "right",
              "bottom",
            ]
          }
          footerContent={null}
          headerContent={
            <Title
              headingLevel="h3"
              size="xl"
            >
              Compliance threshold
            </Title>
          }
          hideOnOutsideClick={true}
          isVisible={null}
          maxWidth="calc(2rem + 18.75rem)"
          onHidden={[Function]}
          onHide={[Function]}
          onMount={[Function]}
          onShow={[Function]}
          onShown={[Function]}
          position="top"
          shouldClose={[Function]}
          tippyProps={Object {}}
          zIndex={9999}
        >
          <OutlinedQuestionCircleIcon
            color="currentColor"
            noVerticalAlign={false}
            size="sm"
            title={null}
          />
        </Popover>
      </React.Fragment>
    }
  >
    <Field
      aria-label="compliance threshold"
      component={[Function]}
      defaultValue={10}
      id="complianceThreshold"
      isRequired={true}
      isValid={true}
      name="complianceThreshold"
      onChange={[Function]}
      style={
        Object {
          "display": "block",
          "width": "60%",
        }
      }
      type="number"
    />
  </FormGroup>
</Fragment>
`;
